src/core/reference/include/openvino/reference/paged_attention.hpp

```diff
@@ -0,0 +1,330 @@
+                                               // {num_rotated_blocks, 1})
+    const Shape& rotation_trig_lut_shape) {    // shape of rotation_trig_lut
+
+    // Use default values for optional parameters if pointers are null.
+    T scale = scale_ptr ? scale_ptr[0] : T(1);
+    int32_t sliding_window = sliding_window_ptr ? sliding_window_ptr[0] : 0;
+    int32_t max_context_len = max_context_len_ptr ? max_context_len_ptr[0] : 0;
+
+    // Determine dimensions.
+    int32_t num_blocks = kv_cache_shape[0];
+    int32_t num_kv_heads = kv_cache_shape[1];
+    int32_t block_size = kv_cache_shape[2];
+    int32_t head_size = kv_cache_shape[3];
+
+    int32_t batch_tokens = q_shape[0];
+    int32_t query_features = q_shape[1];  // equals num_heads * head_size.
+    int32_t num_heads = query_features / head_size;
+    int32_t batch_seq = !past_lens_shape.empty() ? past_lens_shape[0] : 1;
+
+    // Process each query token.
+    for (int32_t token_idx = 0; token_idx < batch_tokens; token_idx++) {
+        // Determine sequence index using subsequence_begins if available.
+        int32_t seq_idx = 0;
+        if (batch_seq > 1 && subsequence_begins) {
+            for (int32_t s = 0; s < batch_seq; s++) {
+                if (token_idx >= subsequence_begins[s] && token_idx < subsequence_begins[s + 1]) {
+                    seq_idx = s;
+                    break;
+                }
+            }
+        }
+        // Process each head.
+        for (int32_t h = 0; h < num_heads; h++) {
+            const T* q_vec = query + token_idx * query_features + h * head_size;
+
+            int32_t seq_new_tokens =
+                subsequence_begins ? (subsequence_begins[seq_idx + 1] - subsequence_begins[seq_idx]) : 0;
+            int32_t seq_past_tokens = past_lens ? past_lens[seq_idx] : 0;
+            int32_t total_keys = seq_past_tokens + seq_new_tokens;
+            std::vector<T> scores(total_keys, T(0));
+
+            // Compute raw attention scores.
+            for (int32_t k = 0; k < total_keys; k++) {
+                T score_val = T(0);
+                if (k < seq_past_tokens) {
+                    // Retrieve key from cache.
+                    int32_t block_id = -1, token_offset = 0;
+                    bool found = paged_attention_utils::find_cached_token(seq_idx,
+                                                                          k,
+                                                                          block_indices,
+                                                                          block_indices_begins,
+                                                                          num_blocks,
+                                                                          block_size,
+                                                                          block_id,
+                                                                          token_offset);
+                    if (!found)
```

Comment:
If not found it should be a critical problem and using an `OPENVINO_ASSERT` may be better for identifying bug?